File: pipelines/artifact_removal/main_run.py
import os
import sys
import pickle
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import welch

# Добавляем корневую директорию проекта в sys.path
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..')))

from models.oscillations.osc_sim import generate_distributed_sources
from models.emg_noise.emg_sim import generate_emg_noise

def main():
    # 1. Загрузка прямой модели (forward model)
    fm_path = os.path.join(os.path.dirname(__file__), '..', '..', 'forward_model', 'forward_model.pkl')
    if not os.path.exists(fm_path):
        print(f"Error: Forward model not found at {fm_path}")
        print("Please run get_fm.py or create a dummy forward_model.pkl first.")
        return

    with open(fm_path, 'rb') as f:
        fm_data = pickle.load(f)

    leadfield = fm_data['leadfield']
    src_pos = fm_data['src_pos']

    # Параметры симуляции
    Fs = 250  # Частота дискретизации (Гц)
    Ts = 10   # Длительность симуляции (сек)
    N = int(Fs * Ts)

    Nsrc = 100       # Общее число мозговых источников
    Ndistr = 10      # Количество "целевых" источников мозга
    flanker = 1      # Фланкер для краевых эффектов (сек)

    N_emg = 20       # Количество источников миографического шума

    # 2. Генерация фонового мозга и целевых осцилляций
    print("Генерация мозговой активности...")
    X_s, X_bg, X_n, z, GA, S = generate_distributed_sources(
        G=leadfield, Nsrc=Nsrc, Ndistr=Ndistr, flanker=flanker, Ts=Ts, Fs=Fs
    )

    # 3. Генерация миографического шума (EMG)
    print("Генерация миографического шума...")
    X_emg, S_emg, GA_emg = generate_emg_noise(
        G=leadfield, src_pos=src_pos, N_emg=N_emg, N=N, Fs=Fs, superficial_ratio=0.8
    )

    # Масштабирование (настройка отношения сигнал/шум)
    # Здесь можно регулировать амплитуду EMG относительно мозга
    emg_multiplier = 2.0
    X_emg = X_emg * emg_multiplier

    white_noise_multiplier = 0.5
    X_n = X_n * white_noise_multiplier

    # 4. Смешивание сигналов
    # Чистый сигнал мозга = Целевые осцилляции + фоновая активность
    X_brain = X_s + X_bg

    # Зашумленный сигнал = Мозг + EMG + белый шум сенсоров
    X_mixed = X_brain + X_emg + X_n

    # 5. Визуализация результатов
    print("Построение визуализаций...")
    time = np.arange(N) / Fs

    # Выбираем один сенсор (например, первый) для визуализации
    sensor_idx = 0

    fig, axs = plt.subplots(3, 2, figsize=(15, 10))

    # --- Временные ряды ---
    axs[0, 0].plot(time, X_brain[sensor_idx, :], color='blue')
    axs[0, 0].set_title("Чистый сигнал мозга (Sensor 0)")
    axs[0, 0].set_xlabel("Time (s)")
    axs[0, 0].set_ylabel("Amplitude")

    axs[1, 0].plot(time, X_emg[sensor_idx, :], color='red')
    axs[1, 0].set_title("Миографический шум (Sensor 0)")
    axs[1, 0].set_xlabel("Time (s)")
    axs[1, 0].set_ylabel("Amplitude")

    axs[2, 0].plot(time, X_mixed[sensor_idx, :], color='purple')
    axs[2, 0].set_title("Смешанный зашумленный сигнал (Sensor 0)")
    axs[2, 0].set_xlabel("Time (s)")
    axs[2, 0].set_ylabel("Amplitude")

    # --- Спектры мощности (PSD) ---
    f, Pxx_brain = welch(X_brain[sensor_idx, :], fs=Fs, nperseg=Fs*2)
    axs[0, 1].semilogy(f, Pxx_brain, color='blue')
    axs[0, 1].set_title("PSD чистого сигнала мозга")
    axs[0, 1].set_xlabel("Frequency (Hz)")
    axs[0, 1].set_ylabel("Power")
    axs[0, 1].set_xlim(0, 100)

    f, Pxx_emg = welch(X_emg[sensor_idx, :], fs=Fs, nperseg=Fs*2)
    axs[1, 1].semilogy(f, Pxx_emg, color='red')
    axs[1, 1].set_title("PSD миографического шума")
    axs[1, 1].set_xlabel("Frequency (Hz)")
    axs[1, 1].set_ylabel("Power")
    axs[1, 1].set_xlim(0, 100)

    f, Pxx_mixed = welch(X_mixed[sensor_idx, :], fs=Fs, nperseg=Fs*2)
    axs[2, 1].semilogy(f, Pxx_mixed, color='purple')
    axs[2, 1].set_title("PSD смешанного сигнала")
    axs[2, 1].set_xlabel("Frequency (Hz)")
    axs[2, 1].set_ylabel("Power")
    axs[2, 1].set_xlim(0, 100)

    plt.tight_layout()
    # Сохраняем график, чтобы не блокировать выполнение
    output_fig = os.path.join(os.path.dirname(__file__), 'simulation_results.png')
    plt.savefig(output_fig)
    print(f"Результаты сохранены в: {output_fig}")
    plt.close(fig)

if __name__ == "__main__":
    main()

File: pipelines/artifact_removal/main_run.m
% Скрипт полного пайплайна для генерации, смешивания и визуализации
% чистого ЭЭГ сигнала и миографического шума.

% Добавляем пути к моделям
addpath('../../models/oscillations');
addpath('../../models/emg_noise');

% 1. Загрузка прямой модели (forward model)
fm_path = '../../forward_model/forward_model.mat';
if ~isfile(fm_path)
    error('Forward model not found at %s. Please run get_fm.py or create dummy model.', fm_path);
end
load(fm_path, 'leadfield', 'src_pos');

% Параметры симуляции
Fs = 250;  % Частота дискретизации (Гц)
Ts = 10;   % Длительность симуляции (сек)
N = Fs * Ts;

Nsrc = 100;       % Общее число мозговых источников
Ndistr = 10;      % Количество "целевых" источников мозга
flanker = 1;      % Фланкер для краевых эффектов (сек)

N_emg = 20;       % Количество источников миографического шума

% 2. Генерация фонового мозга и целевых осцилляций
disp('Генерация мозговой активности...');
[X_s, X_bg, X_n, z, GA, S] = osc_sim(leadfield, Nsrc, Ndistr, flanker, Ts, Fs);

% 3. Генерация миографического шума (EMG)
disp('Генерация миографического шума...');
[X_emg, S_emg, GA_emg] = emg_sim(leadfield, src_pos, N_emg, N, Fs, 0.8);

% Масштабирование (настройка отношения сигнал/шум)
emg_multiplier = 2.0;
X_emg = X_emg * emg_multiplier;

white_noise_multiplier = 0.5;
X_n = X_n * white_noise_multiplier;

% 4. Смешивание сигналов
% Чистый сигнал мозга = Целевые осцилляции + фоновая активность
X_brain = X_s + X_bg;

% Зашумленный сигнал = Мозг + EMG + белый шум сенсоров
X_mixed = X_brain + X_emg + X_n;

% 5. Визуализация результатов
disp('Построение визуализаций...');
time = (0:N-1) / Fs;

% Выбираем один сенсор (например, первый) для визуализации
sensor_idx = 1;

figure('Position', [100, 100, 1200, 800]);

% --- Временные ряды ---
subplot(3, 2, 1);
plot(time, X_brain(sensor_idx, :), 'b');
title('Чистый сигнал мозга (Sensor 1)');
xlabel('Time (s)'); ylabel('Amplitude');

subplot(3, 2, 3);
plot(time, X_emg(sensor_idx, :), 'r');
title('Миографический шум (Sensor 1)');
xlabel('Time (s)'); ylabel('Amplitude');

subplot(3, 2, 5);
plot(time, X_mixed(sensor_idx, :), 'Color', [0.5, 0, 0.5]);
title('Смешанный зашумленный сигнал (Sensor 1)');
xlabel('Time (s)'); ylabel('Amplitude');

% --- Спектры мощности (PSD) ---
% Используем pwelch для оценки спектра
window = round(Fs * 2);

[Pxx_brain, f] = pwelch(X_brain(sensor_idx, :), window, [], [], Fs);
subplot(3, 2, 2);
semilogy(f, Pxx_brain, 'b');
title('PSD чистого сигнала мозга');
xlabel('Frequency (Hz)'); ylabel('Power');
xlim([0 100]);

[Pxx_emg, f] = pwelch(X_emg(sensor_idx, :), window, [], [], Fs);
subplot(3, 2, 4);
semilogy(f, Pxx_emg, 'r');
title('PSD миографического шума');
xlabel('Frequency (Hz)'); ylabel('Power');
xlim([0 100]);

[Pxx_mixed, f] = pwelch(X_mixed(sensor_idx, :), window, [], [], Fs);
subplot(3, 2, 6);
semilogy(f, Pxx_mixed, 'Color', [0.5, 0, 0.5]);
title('PSD смешанного сигнала');
xlabel('Frequency (Hz)'); ylabel('Power');
xlim([0 100]);

% Сохраняем график
saveas(gcf, 'simulation_results_matlab.png');
disp('Результаты сохранены в: simulation_results_matlab.png');
